services/storage-users/pkg/command/trash_bin.go

```diff
@@ -53,3 +89,402 @@ func PurgeExpiredResources(cfg *config.Config) *cli.Command {
+			defer span.End()
+
+			res, err := client.ListRecycle(ctx, &provider.ListRecycleRequest{Ref: &ref, Key: "/"})
+			if err != nil {
+				log.Error().Err(err).Msg(retrievingErrorMsg)
+				return err
+			}
+			if res.Status.Code != rpc.Code_CODE_OK {
+				return fmt.Errorf("%s %s", retrievingErrorMsg, res.Status.Code)
+			}
+
+			var found bool
+			var itemRef *provider.RecycleItem
+			for _, item := range res.GetRecycleItems() {
+				if item.GetKey() == itemID {
+					itemRef = item
+					found = true
+					break
+				}
+			}
+			if !found {
+				return fmt.Errorf("itemID '%s' not found", itemID)
+			}
+			fmt.Printf("\nRun restoring with option=%s\n", optionFlagVal)
+			fmt.Printf("restoring itemID: '%s', path: '%s', type: '%s'\n", itemRef.GetKey(), itemRef.GetRef().GetPath(), itemType(itemRef.GetType()))
+			dstRes, err := restore(ctx, client, ref, itemRef, overwriteOption)
+			if err != nil {
+				return err
+			}
+			fmt.Printf("itemID: '%s', path: '%s', restored as '%s'\n", itemRef.GetKey(), itemRef.GetRef().GetPath(), dstRes.GetPath())
+			return nil
+		},
+	}
+}
+
+func restore(ctx context.Context, client gateway.GatewayAPIClient, ref provider.Reference, item *provider.RecycleItem, overwriteOption int) (*provider.Reference, error) {
+	dst, _ := deepcopy.Copy(ref).(provider.Reference)
+	dst.Path = utils.MakeRelativePath(item.GetRef().GetPath())
+	// Restore request
+	req := &provider.RestoreRecycleItemRequest{
+		Ref:        &ref,
+		Key:        path.Join(item.GetKey(), "/"),
+		RestoreRef: &dst,
+	}
+
+	exists, dstStatRes, err := isDestinationExists(ctx, client, dst)
+	if err != nil {
+		return &dst, err
+	}
+
+	if exists {
+		fmt.Printf("destination '%s' exists.\n", dstStatRes.GetInfo().GetPath())
+		switch overwriteOption {
+		case SKIP:
+			return &dst, nil
+		case REPLACE:
+			// delete existing tree
+			delReq := &provider.DeleteRequest{Ref: &dst}
+			delRes, err := client.Delete(ctx, delReq)
+			if err != nil {
+				return &dst, fmt.Errorf("error sending grpc delete request %w", err)
+			}
+			if delRes.Status.Code != rpc.Code_CODE_OK && delRes.Status.Code != rpc.Code_CODE_NOT_FOUND {
+				return &dst, fmt.Errorf("deleting error %w", err)
+			}
+		case KEEP_BOTH:
+			// modify the file name
+			req.RestoreRef, err = resolveDestination(ctx, client, dst)
+			if err != nil {
+				return &dst, fmt.Errorf("trash-bin item restoring error %w", err)
+			}
+		}
+	}
+
+	res, err := client.RestoreRecycleItem(ctx, req)
+	if err != nil {
+		log.Error().Err(err).Msg("trash-bin item restoring error")
+		return req.RestoreRef, err
+	}
+	if res.Status.Code != rpc.Code_CODE_OK {
+		return req.RestoreRef, fmt.Errorf("trash-bin item restoring error %s", res.Status.Code)
+	}
+	return req.RestoreRef, nil
+}
+
+func resolveDestination(ctx context.Context, client gateway.GatewayAPIClient, dstRef provider.Reference) (*provider.Reference, error) {
+	dst := dstRef
+	for i := 1; i < 100; i++ {
```

Comment:
Maybe extract the `100` to a global var? I could imagine scenarios where might want to increase this limit easily :wink: